include/crm/common/xml_internal.h

```diff
@@ -0,0 +1,118 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+#ifndef CRM_COMMON_XML_INTERNAL__H
+#  define CRM_COMMON_XML_INTERNAL__H
+
+/**
+ * \file
+ * \brief Internal-only wrappers for and extensions to libxml2 (libxslt)
+ * \ingroup core
+ */
+
+#  include <stdlib.h>
+#  include <stdio.h>
+#  include <string.h>
+
+#  include <crm/crm.h>  /* transitively imports qblog.h */
+
+
+/*!
+ * \brief Base for directing lib{xml2,xslt} log into standard libqb backend
+ *
+ * This macro implements the core of what can be needed for directing
+ * libxml2 or libxslt error messaging into standard, preconfigured
+ * libqb-backed log stream.
+ *
+ * It's a bit unfortunate that libxml2 (and more sparsely, also libxslt)
+ * emits a single message by chunks (location is emitted separatedly from
+ * the * message itself), so we have to take the effort to combine these
+ * chunks back to single message.  Whether to do this or not is driven
+ * with \p dechunk toggle.
+ *
+ * The form of a macro was chosen for implicit deriving of __FILE__, etc.
+ * and also because static dechunking buffer should be differentiated per
+ * library (here we assume different functions referring to this macro
+ * will not ever be using both at once), preferably also per-library
+ * context of use to avoid clashes altogether.
+ *
+ * Note that we cannot use qb_logt, because callsite data have to be known
+ * at the moment of compilation, which it is not always the case -- xml_log
+ * (and unfortunately there's no clear explanation of the fail to compile).
+ *
+ * \param[in] priority Syslog priority for the message to be logged
+ * \param[in] dechunk  Whether to dechunk new-line terminated message
+ * \param[in] postemit Code to be executed once message is sent out
+ * \param[in] prefix   How to prefix the message or NULL for raw passing
+ * \param[in] fmt      Format string as with printf-like functions
+ * \param[in] ap       Variable argument list to supplement \p fmt format string
+ */
+#define CRM_XML_LOG_BASE(priority, dechunk, postemit, prefix, fmt, ap)        \
+do {                                                                          \
+    if (!(dechunk) && (prefix) == NULL) {  /* quick pass */                   \
+        qb_log_from_external_source_va(__FUNCTION__, __FILE__, (fmt),         \
+                                       (priority), __LINE__, 0, (ap));        \
+        (void) (postemit);                                                    \
+    } else {                                                                  \
+        int len = 0;                                                          \
+        char *buf = NULL;                                                     \
+        static int buffer_len = 0;                                            \
+        static char *buffer = NULL;                                           \
```

Comment:
Well, we currently does not enable `-Wshadow` compilation flag, likely because
there are some clashes yet to be solved.  But as demonstrated here, it is pretty
useful check so I'll take a look into enabling that in a separate PR.